File: dexter-ts/src/utils/context.ts
import { existsSync, readFileSync, writeFileSync, mkdirSync } from 'fs';
import { join } from 'path';
import { createHash } from 'crypto';
import { callLlm, DEFAULT_MODEL } from '../model/llm.js';
import { DEFAULT_SYSTEM_PROMPT, CONTEXT_SELECTION_SYSTEM_PROMPT } from '../agent/prompts.js';
import { SelectedContextsSchema } from '../agent/schemas.js';

interface ContextPointer {
  filepath: string;
  filename: string;
  toolName: string;
  args: Record<string, unknown>;
  summary: string;
  taskId?: number;
  queryId?: string;
}

interface ContextData {
  tool_name: string;
  args: Record<string, unknown>;
  summary: string;
  timestamp: string;
  task_id?: number;
  query_id?: string;
  result: unknown;
}

export class ContextManager {
  private contextDir: string;
  private model: string;
  public pointers: ContextPointer[] = [];

  constructor(contextDir: string = '.dexter/context', model: string = DEFAULT_MODEL) {
    this.contextDir = contextDir;
    this.model = model;
    if (!existsSync(contextDir)) {
      mkdirSync(contextDir, { recursive: true });
    }
  }

  private hashArgs(args: Record<string, unknown>): string {
    const argsStr = JSON.stringify(args, Object.keys(args).sort());
    return createHash('md5').update(argsStr).digest('hex').slice(0, 12);
  }

  hashQuery(query: string): string {
    return createHash('md5').update(query).digest('hex').slice(0, 12);
  }

  private generateFilename(toolName: string, args: Record<string, unknown>): string {
    const argsHash = this.hashArgs(args);
    const ticker = typeof args.ticker === 'string' ? args.ticker.toUpperCase() : null;
    return ticker 
      ? `${ticker}_${toolName}_${argsHash}.json`
      : `${toolName}_${argsHash}.json`;
  }

  private async generateSummary(
    toolName: string,
    args: Record<string, unknown>,
    result: unknown
  ): Promise<string> {
    const resultStr = JSON.stringify(result).slice(0, 1000);

    const prompt = `
    Tool: ${toolName}
    Arguments: ${JSON.stringify(args, null, 2)}
    Output preview: ${resultStr}
    
    Generate a brief one-sentence summary describing what data this tool output contains.
    Focus on the key information (e.g., "Apple's last 4 quarterly income statements from Q1 2023 to Q4 2023").
    `;

    try {
      const response = await callLlm(prompt, {
        systemPrompt: DEFAULT_SYSTEM_PROMPT,
        model: this.model,
      });
      return typeof response === 'string' ? response.trim() : String(response).trim();
    } catch {
      return `${toolName} output with args ${JSON.stringify(args)}`;
    }
  }

  async saveContext(
    toolName: string,
    args: Record<string, unknown>,
    result: unknown,
    taskId?: number,
    queryId?: string
  ): Promise<string> {
    const filename = this.generateFilename(toolName, args);
    const filepath = join(this.contextDir, filename);

    const summary = await this.generateSummary(toolName, args, result);

    const contextData: ContextData = {
      tool_name: toolName,
      args: args,
      summary: summary,
      timestamp: new Date().toISOString(),
      task_id: taskId,
      query_id: queryId,
      result: result,
    };

    writeFileSync(filepath, JSON.stringify(contextData, null, 2));

    const pointer: ContextPointer = {
      filepath,
      filename,
      toolName,
      args,
      summary,
      taskId,
      queryId,
    };

    this.pointers.push(pointer);

    return filepath;
  }

  getAllPointers(): ContextPointer[] {
    return [...this.pointers];
  }

  getPointersForQuery(queryId: string): ContextPointer[] {
    return this.pointers.filter(p => p.queryId === queryId);
  }

  loadContexts(filepaths: string[]): ContextData[] {
    const contexts: ContextData[] = [];
    for (const filepath of filepaths) {
      try {
        const content = readFileSync(filepath, 'utf-8');
        contexts.push(JSON.parse(content));
      } catch (e) {
        console.warn(`Warning: Failed to load context file ${filepath}: ${e}`);
      }
    }
    return contexts;
  }

  async selectRelevantContexts(
    query: string,
    availablePointers: ContextPointer[]
  ): Promise<string[]> {
    if (availablePointers.length === 0) {
      return [];
    }

    const pointersInfo = availablePointers.map((ptr, i) => ({
      id: i,
      tool_name: ptr.toolName,
      args: ptr.args,
      summary: ptr.summary,
    }));

    const prompt = `
    Original user query: "${query}"
    
    Available tool outputs:
    ${JSON.stringify(pointersInfo, null, 2)}
    
    Select which tool outputs are relevant for answering the query.
    Return a JSON object with a "context_ids" field containing a list of IDs (0-indexed) of the relevant outputs.
    Only select outputs that contain data directly relevant to answering the query.
    `;

    try {
      const response = await callLlm(prompt, {
        systemPrompt: CONTEXT_SELECTION_SYSTEM_PROMPT,
        model: this.model,
        outputSchema: SelectedContextsSchema,
      });

      const selectedIds = (response as { context_ids: number[] }).context_ids || [];

      return selectedIds
        .filter((idx) => idx >= 0 && idx < availablePointers.length)
        .map((idx) => availablePointers[idx].filepath);
    } catch (e) {
      console.warn(`Warning: Context selection failed: ${e}, loading all contexts`);
      return availablePointers.map((ptr) => ptr.filepath);
    }
  }
}

File: legacy/src/dexter/utils/context.py
import json
import hashlib
import os
from datetime import datetime
from typing import List, Dict, Any, Optional
from pathlib import Path
from pydantic import BaseModel

from dexter.model import call_llm, DEFAULT_MODEL
from dexter.prompts import DEFAULT_SYSTEM_PROMPT, CONTEXT_SELECTION_SYSTEM_PROMPT


class ContextManager:
    """Manages context offloading and onloading for tool outputs."""
    
    def __init__(self, context_dir: str = ".dexter/context", model: str = DEFAULT_MODEL):
        """
        Initialize the context manager.
        
        Args:
            context_dir: Directory path for storing context files
            model: The model to use for LLM calls
        """
        self.context_dir = Path(context_dir)
        self.context_dir.mkdir(parents=True, exist_ok=True)
        self.pointers: List[Dict[str, Any]] = []
        self.model = model
    
    def _hash_args(self, args: dict) -> str:
        """Generate a hash of tool arguments for filename."""
        # Sort args to ensure consistent hashing
        args_str = json.dumps(args, sort_keys=True, default=self._json_serializer)
        return hashlib.md5(args_str.encode()).hexdigest()[:12]
    
    def _json_serializer(self, obj):
        """Custom JSON serializer for Pydantic models and other special types."""
        if isinstance(obj, BaseModel):
            # Use model_dump() for Pydantic v2, fallback to dict() for v1
            if hasattr(obj, 'model_dump'):
                return obj.model_dump()
            elif hasattr(obj, 'dict'):
                return obj.dict()
            else:
                return str(obj)
        elif isinstance(obj, datetime):
            return obj.isoformat()
        elif isinstance(obj, (list, tuple)):
            return [self._json_serializer(item) for item in obj]
        elif isinstance(obj, dict):
            return {k: self._json_serializer(v) for k, v in obj.items()}
        # Let JSON handle primitive types (str, int, float, bool, None)
        return obj
    
    def _generate_summary(self, tool_name: str, args: dict, result: Any) -> str:
        """Generate a brief summary snippet of the tool output using LLM."""
        # Convert result to string representation for the prompt
        result_str = json.dumps(result, default=str)[:1000]  # Limit size for prompt
        
        prompt = f"""
        Tool: {tool_name}
        Arguments: {json.dumps(args, indent=2)}
        Output preview: {result_str}
        
        Generate a brief one-sentence summary describing what data this tool output contains.
        Focus on the key information (e.g., "Apple's last 4 quarterly income statements from Q1 2023 to Q4 2023").
        """
        
        try:
            response = call_llm(
                prompt,
                system_prompt=DEFAULT_SYSTEM_PROMPT,
                model=self.model
            )
            summary = response.content if hasattr(response, 'content') else str(response)
            return summary.strip()
        except Exception as e:
            # Fallback to a simple description if LLM fails
            return f"{tool_name} output with args {args}"
    
    def save_context(
        self,
        tool_name: str,
        args: dict,
        result: Any,
        task_id: Optional[int] = None
    ) -> str:
        """
        Save tool output to a JSON file and return the file path.
        
        Args:
            tool_name: Name of the tool
            args: Tool arguments used
            result: Tool output result
            task_id: Optional task ID associated with this output
            
        Returns:
            File path to the saved context
        """
        # Generate filename
        args_hash = self._hash_args(args)
        filename = f"{tool_name}_{args_hash}.json"
        filepath = self.context_dir / filename
        
        # Generate summary
        summary = self._generate_summary(tool_name, args, result)
        
        # Prepare context data (metadata first, result last)
        context_data = {
            "tool_name": tool_name,
            "args": args,
            "summary": summary,
            "timestamp": datetime.now().isoformat(),
            "task_id": task_id,
            "result": result
        }
        
        # Save to file with custom serializer for Pydantic models
        with open(filepath, 'w') as f:
            json.dump(context_data, f, default=self._json_serializer, indent=2)
        
        # Create pointer metadata
        pointer = {
            "filepath": str(filepath),
            "filename": filename,
            "tool_name": tool_name,
            "args": args,
            "summary": summary,
            "task_id": task_id
        }
        
        self.pointers.append(pointer)
        
        return str(filepath)
    
    def get_all_pointers(self) -> List[Dict[str, Any]]:
        """Return all tracked pointer metadata."""
        return self.pointers.copy()
    
    def load_contexts(self, filepaths: List[str]) -> List[Dict[str, Any]]:
        """
        Load context files and return their data.
        
        Args:
            filepaths: List of file paths to load
            
        Returns:
            List of context data dictionaries
        """
        contexts = []
        for filepath in filepaths:
            try:
                with open(filepath, 'r') as f:
                    context_data = json.load(f)
                    contexts.append(context_data)
            except Exception as e:
                # Log error but continue loading other files
                print(f"Warning: Failed to load context file {filepath}: {e}")
        
        return contexts
    
    def select_relevant_contexts(
        self,
        query: str,
        available_pointers: List[Dict[str, Any]]
    ) -> List[str]:
        """
        Use LLM to select which context files are relevant for answering the query.
        
        Args:
            query: Original user query
            available_pointers: List of pointer metadata dictionaries
            
        Returns:
            List of selected file paths
        """
        if not available_pointers:
            return []
        
        # Format pointers for the prompt
        pointers_info = []
        for i, ptr in enumerate(available_pointers):
            pointers_info.append({
                "id": i,
                "tool_name": ptr["tool_name"],
                "args": ptr["args"],
                "summary": ptr["summary"]
            })
        
        prompt = f"""
        Original user query: "{query}"
        
        Available tool outputs:
        {json.dumps(pointers_info, indent=2, default=self._json_serializer)}
        
        Select which tool outputs are relevant for answering the query.
        Return a JSON object with a "context_ids" field containing a list of IDs (0-indexed) of the relevant outputs.
        Only select outputs that contain data directly relevant to answering the query.
        """
        
        from dexter.schemas import SelectedContexts
        
        try:
            response = call_llm(
                prompt,
                system_prompt=CONTEXT_SELECTION_SYSTEM_PROMPT,
                output_schema=SelectedContexts,
                model=self.model
            )
            
            # Extract selected IDs
            if isinstance(response, SelectedContexts):
                selected_ids = response.context_ids
            elif hasattr(response, 'context_ids'):
                selected_ids = response.context_ids
            else:
                # Fallback: return all if selection fails
                selected_ids = list(range(len(available_pointers)))
            
            # Map IDs to filepaths
            selected_filepaths = []
            for idx in selected_ids:
                if 0 <= idx < len(available_pointers):
                    selected_filepaths.append(available_pointers[idx]["filepath"])
            
            return selected_filepaths
        except Exception as e:
            # Fallback: return all contexts if selection fails
            print(f"Warning: Context selection failed: {e}, loading all contexts")
            return [ptr["filepath"] for ptr in available_pointers]

